Kll non sampled

pengzai6666 opened the pull request:
*Issue #, if available:*
This version hasn't been verified on whether the integration of KLL Sketch is successful or not by unit testing, and there may be ways to optimize the functions related to the integration. Moreover, I implement the KLLSketch class by directly extending the Analyzer class, instead of using the UserDefinedAggregateFunction, which Sebastian suggests as a possible pointer to do the integration. I would greatly appreciate it if I could know if this way of integration is acceptable.

*Description of changes:*
Integrate KLL Sketch (without a Sampler) into Deequ. 

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

sscdotopen commented on README.md:
```diff
@@ -17,13 +17,13 @@ __Maven__
 <dependency>
   <groupId>com.amazon.deequ</groupId>
   <artifactId>deequ</artifactId>
-  <version>1.0.2</version>
```
Please rebase to the latest version

pengzai6666 replied:
Sure!

sscdotopen commented on README.md:
````diff
@@ -17,13 +17,13 @@ __Maven__
 __sbt__
 ```
-libraryDependencies += "com.amazon.deequ" % "deequ" % "1.0.2"
+libraryDependencies += "com.amazon.deequ" % "deequ" % "1.0.1"
````
Please rebase to the latest version

pengzai6666 replied:
Sure!

sscdotopen commented on pom.xml:
```diff
@@ -116,6 +116,12 @@
             <scope>test</scope>
         </dependency>
 
+        <dependency>
```
Why is this dependecy required?

pengzai6666 replied:
This dependency was added to enable Spark UI (for testing purpose), now get it removed.

pengzai6666 commented:
Hi Sebastian,

Thank you very much for your feedback on my PR. Actually the reason that we take the route of implementing Analyzer in this PR is because we have tried to use the UserDefinedAggregateFunction pointer that you suggests before, but found out that we can't use ObjectType in the bufferSchema, so we cannot pass a KLL Object to the buffer and use it in update() or merge() method. Therefore now what we are trying to do is to serialize the KLL Object to JSON String, pass it to MutableAggregationBuffer, and deserialize the JSON string to object every time when it is needed in update() and merge() method. I am a little concerned that this route may cause performance issues as well. I would greatly appreciate it if you could share your opinion on this route.

sscdotopen commented:
Hi Peng,
I think the route which you are proposing will not work. If you have a dataframe with 100 columns and you want to compute sketches for all columns, you will need to run 100 Spark jobs instead of one. If the DataFrame is not cached, this can seriously impact performance.

You could try to extend the KLLSketch to be able to serialize itself to and from a byte array, and integrate it this way. This might negatively impact performance as well, but not as much as having to run a large number of extra jobs.

sscdotopen commented on pom.xml:
```diff
@@ -115,7 +115,7 @@
             <scope>test</scope>
         </dependency>
-
+        
```
please remove the empty line here

pengzai6666 replied:
Sure!

sscdotopen commented on src/main/scala/com/amazon/deequ/analyzers/KLLSketch.scala:
```diff
@@ -0,0 +1,122 @@
+package com.amazon.deequ.analyzers
```
This file needs a license header

pengzai6666 replied:
Sure.

sscdotopen commented on src/main/scala/com/amazon/deequ/analyzers/KLLSketch.scala:
```diff
@@ -0,0 +1,122 @@
+import scala.util.Failure
+
+/** State for KLL Sketches */
+case class KLLState(qSketch: QuantileNonSample[Double], globalMax: Double, globalMin: Double)
```
Please format according to our [coding conventions](https://github.com/databricks/scala-style-guide)

pengzai6666 replied:
Yes I have read it and try to format my code according to conventions. But I could miss out/ misunderstand something, therefore please take a second look after I make the next commit.

sscdotopen commented on src/main/scala/com/amazon/deequ/analyzers/KLLSketch.scala:
```diff
@@ -0,0 +1,122 @@
+}
+
+object KLLState{
+  // from Array[Byte] to original type
```
Please use /**/ comments starting with capital letters

pengzai6666 replied:
Sure. Add a /**/ comments on the fromBytes function.

sscdotopen commented on src/main/scala/com/amazon/deequ/analyzers/KLLSketch.scala:
```diff
@@ -0,0 +1,122 @@
+    val max = buffer.getDouble
+    val store = new Array[Byte](buffer.remaining())
+    buffer.get(store)
+    val obj = KLLSketchSerializer.serializer.deserialize(store)
```
please use a more descriptive variable name

pengzai6666 replied:
Done. Use kllBuffer and kllSketch to replace store and obj respectively.

sscdotopen commented on src/main/scala/com/amazon/deequ/analyzers/KLLSketch.scala:
```diff
@@ -0,0 +1,122 @@
+}
+
+/**
+ *
```
Please add a comment here

pengzai6666 replied:
Done.

sscdotopen commented on src/main/scala/com/amazon/deequ/analyzers/KLLSketch.scala:
```diff
@@ -0,0 +1,122 @@
+ *
+ * @param column Which column to compute this aggregation on.
+ */
+case class KLLSketch(column: String, where: Option[String] = None, sketchSize: Int = KLLSketch.defaultSketchSize,
```
Please format according to our [coding conventions](https://github.com/databricks/scala-style-guide)

pengzai6666 replied:
Done.

sscdotopen commented on src/main/scala/com/amazon/deequ/analyzers/KLLSketch.scala:
```diff
@@ -0,0 +1,122 @@
+  extends ScanShareableAnalyzer[KLLState, KLLMetric] {
+
+  private[this] val PARAM_CHECK: StructType => Unit = { _ =>
+        if (maxDetailBins > KLLSketch.MaximumAllowedDetailBins) {
```
please use 2 space indentation

pengzai6666 replied:
Done.

sscdotopen commented on src/main/scala/com/amazon/deequ/analyzers/KLLSketch.scala:
```diff
@@ -0,0 +1,122 @@
+      }
+
+  override def aggregationFunctions(): Seq[Column] = {
+    stateful_kll(conditionalSelection(column, where),sketchSize, shrinkingFactor) :: Nil
```
space before sketchSize

pengzai6666 replied:
Sure.

sscdotopen commented on src/main/scala/com/amazon/deequ/analyzers/KLLSketch.scala:
```diff
@@ -0,0 +1,122 @@
+                BucketValue(lowBound,highBound,finalSketch.getRank(highBound) -
+                finalSketch.getRankExclusive(lowBound))
+            } else {
+              bucketsList += BucketValue(lowBound,highBound,finalSketch.getRankExclusive(highBound) -
```
please use spaces after , and move the expression `finalSketch.getRankExclusive(highBound) - finalSketch.getRankExclusive(lowBound)` to its own line to simplify the code

pengzai6666 replied:
Done. Great point!

sscdotopen commented on src/main/scala/com/amazon/deequ/analyzers/NonSampleCompactor.scala:
```diff
@@ -0,0 +1,47 @@
+
+/**
+ * A quantile sketcher whose output is half the size of its input
+ *
```
please remove the empty line here

pengzai6666 replied:
Done.

sscdotopen commented on src/main/scala/com/amazon/deequ/analyzers/NonSampleCompactor.scala:
```diff
@@ -0,0 +1,47 @@
+  var numOfCompress = 0
+  var offset = 0
+  var buffer: ArrayBuffer[T] = ArrayBuffer[T]()
+  private def findOdd(items:Int): Option[T] = (items%2) match {
```
please add an empty line before the function definition and spaces around %

pengzai6666 replied:
Sure.

sscdotopen commented on src/main/scala/com/amazon/deequ/analyzers/NonSampleCompactor.scala:
```diff
@@ -0,0 +1,47 @@
+  var offset = 0
+  var buffer: ArrayBuffer[T] = ArrayBuffer[T]()
+  private def findOdd(items:Int): Option[T] = (items%2) match {
+    case 1 => Some(buffer(math.max(items-1,0)))
```
please reformat as `items -1, 0`

pengzai6666 replied:
Done.

sscdotopen commented on src/main/scala/com/amazon/deequ/analyzers/NonSampleCompactor.scala:
```diff
@@ -0,0 +1,47 @@
+  var buffer: ArrayBuffer[T] = ArrayBuffer[T]()
+  private def findOdd(items:Int): Option[T] = (items%2) match {
+    case 1 => Some(buffer(math.max(items-1,0)))
+    case 0 => None
```
please use case _

pengzai6666 replied:
Done.

sscdotopen commented on src/main/scala/com/amazon/deequ/analyzers/NonSampleCompactor.scala:
```diff
@@ -0,0 +1,47 @@
+    if (numOfCompress % 2 == 1) {
+      offset = 1 - offset
+    }
+    // Comment it For unit testing, should uncomment it for real use.
```
Please remove this

pengzai6666 replied:
Done. This is the comment made when developing, sorry for keeping it when making commits.

sscdotopen commented on src/main/scala/com/amazon/deequ/analyzers/NonSampleCompactor.scala:
```diff
@@ -0,0 +1,47 @@
+//    else {
+//      offset = if (Random.nextBoolean()) 1 else 0
+//    }
+    val sBuff = buffer.toArray.slice(0,len).sorted
```
please give this a descriptive variable name and add a space behind the `,`

pengzai6666 replied:
Done. Replace sBuffer with sortedBuffer.

sscdotopen commented on src/main/scala/com/amazon/deequ/analyzers/NonSampleCompactor.scala:
```diff
@@ -0,0 +1,47 @@
+//      offset = if (Random.nextBoolean()) 1 else 0
+//    }
+    val sBuff = buffer.toArray.slice(0,len).sorted
+    val output = (offset until len by 2).map(sBuff(_)).toArray
```
this line of code is very hard to read? what does it do? please rewrite or add a comment

pengzai6666 replied:
It selects half of the items from this level compactor to the next level compactor. 
e.g. if sortedBuffer is Array(1,2,3,4), if offset is 1, output = Array(2,4), and if offset is 0, output = Array(1,3), this will be the input to the next level compactor.

I also add this description and example to the comment in the code to avoid confusion.

sscdotopen commented on src/main/scala/com/amazon/deequ/analyzers/NonSampleCompactor.scala:
```diff
@@ -0,0 +1,47 @@
+    val tail = findOdd(items)
+    items = items % 2
+    var newBuffer = ArrayBuffer[T]()
+    if (tail.isDefined) newBuffer = newBuffer :+ tail.get
```
please add braes for the if statement

pengzai6666 replied:
Done.

sscdotopen commented on src/main/scala/com/amazon/deequ/analyzers/QuantileNonSample.scala:
```diff
@@ -0,0 +1,307 @@
+package com.amazon.deequ.analyzers
```
This file needs a license header

pengzai6666 replied:
Added.

sscdotopen commented on src/main/scala/com/amazon/deequ/analyzers/QuantileNonSample.scala:
```diff
@@ -0,0 +1,307 @@
+import scala.util.control.Breaks._
+
+
+class QuantileNonSample[T](val sketchSize: Int,
```
Please format according to our [coding conventions](https://github.com/databricks/scala-style-guide)

pengzai6666 replied:
Done. Reformat the locations of both variables and extends.

sscdotopen commented on src/main/scala/com/amazon/deequ/analyzers/QuantileNonSample.scala:
```diff
@@ -0,0 +1,307 @@
+                           ct: ClassTag[T]) extends Serializable{
+
+  private var curNumOfCompactors = 0
+  // number of items in compactors
```
Please use /**/ and have comments start with capital letters

pengzai6666 replied:
Done.

sscdotopen commented on src/main/scala/com/amazon/deequ/analyzers/QuantileNonSample.scala:
```diff
@@ -0,0 +1,307 @@
+  private var compactors = ArrayBuffer[NonSampleCompactor[T]]()
+  expand()
+
+  // get and set method for the serialize/deserialize functions
```
Please use /**/ and have comments start with capital letters

pengzai6666 replied:
Added.

sscdotopen commented on src/main/scala/com/amazon/deequ/analyzers/QuantileNonSample.scala:
```diff
@@ -0,0 +1,307 @@
+  private var compactorTotalSize = 0
+  // initialize with ArrayBuffer, add compactors later
+  private var compactors = ArrayBuffer[NonSampleCompactor[T]]()
+  expand()
```
please add an empty line

pengzai6666 replied:
Done.

sscdotopen commented on src/main/scala/com/amazon/deequ/analyzers/QuantileNonSample.scala:
```diff
@@ -0,0 +1,307 @@
+  expand()
+
+  // get and set method for the serialize/deserialize functions
+  def getShrinkingFactor: Double = {
```
Why do we need the accessors here? We can directly access the variables, it does not make a difference

pengzai6666 replied:
Actually what I was thinking is that we should set the variables to private (except sketchSize) to be more secure. If the get/set accessors are public, then there's no difference between setting variables to public and remove accessors. However, maybe later we could add stricter access modifier to these accessors. In that case it would make a difference. But this is only my point so feel free to give your advice on this.

sscdotopen replied:
You can do that if you feel that it is necessary, but IMHO, this Java Bean pattern is rarely used in scala.

pengzai6666 replied:
After a second thought and discuss with Arun, we decide to remove the get/set methods! Thank you for your great advice!

sscdotopen commented on src/main/scala/com/amazon/deequ/analyzers/QuantileNonSample.scala:
```diff
@@ -0,0 +1,307 @@
+
+  def getCompactorItems: Array[Array[T]] = {
+    var ret = ArrayBuffer[Array[T]]()
+    compactors.toArray.foreach {compactor =>
```
please add a space after {

pengzai6666 replied:
Added.

sscdotopen commented on src/main/scala/com/amazon/deequ/analyzers/QuantileNonSample.scala:
```diff
@@ -0,0 +1,307 @@
+    curNumOfCompactors = input
+  }
+
+  def setCompactorActualSize(input : Int): Unit = {
```
Why do we need these accessors here?

pengzai6666 replied:
Removed.

sscdotopen commented on src/main/scala/com/amazon/deequ/analyzers/QuantileNonSample.scala:
```diff
@@ -0,0 +1,307 @@
+    compactorTotalSize = getCompactorCapacityCount
+  }
+
+  private def capacity(height:Int): Int = {
```
please use a space after `:`

pengzai6666 replied:
Added.

sscdotopen commented on src/main/scala/com/amazon/deequ/analyzers/QuantileNonSample.scala:
```diff
@@ -0,0 +1,307 @@
+  }
+
+  /**
+   * update the sketch with a single item
```
Update

pengzai6666 replied:
Updated :)

sscdotopen commented on src/main/scala/com/amazon/deequ/analyzers/QuantileNonSample.scala:
```diff
@@ -0,0 +1,307 @@
+    breakable {
+      for (height <- compactors.indices) {
+        if (compactors(height).buffer.length >= capacity(height)) {
+          if (height + 1 >= curNumOfCompactors) expand()
```
please add braces around the if

pengzai6666 replied:
Added.

sscdotopen commented on src/main/scala/com/amazon/deequ/analyzers/QuantileNonSample.scala:
```diff
@@ -0,0 +1,307 @@
+        if (compactors(height).buffer.length >= capacity(height)) {
+          if (height + 1 >= curNumOfCompactors) expand()
+          val output: Array[T] = compactors(height).compact
+          output.foreach(element => compactors(height + 1).buffer = compactors(height + 1).buffer :+ element)
```
please use `.foreach { }`

pengzai6666 replied:
Updated.

sscdotopen commented on src/main/scala/com/amazon/deequ/analyzers/QuantileNonSample.scala:
```diff
@@ -0,0 +1,307 @@
+    }
+  }
+
+  private def compareGreater[T](o1: T, o2: T)(implicit ord: Ordering[T]): Boolean = ord.gt(o1, o2)
```
please add braces

pengzai6666 replied:
Added.

sscdotopen commented on src/main/scala/com/amazon/deequ/analyzers/QuantileNonSample.scala:
```diff
@@ -0,0 +1,307 @@
+  }
+
+  private def compareGreater[T](o1: T, o2: T)(implicit ord: Ordering[T]): Boolean = ord.gt(o1, o2)
+  private def compareSmaller[T](o1: T, o2: T)(implicit ord: Ordering[T]): Boolean = ord.lt(o1,o2)
```
please add braces

pengzai6666 replied:
Added.

sscdotopen commented on src/main/scala/com/amazon/deequ/analyzers/QuantileNonSample.scala:
```diff
@@ -0,0 +1,307 @@
+  private def compareSmaller[T](o1: T, o2: T)(implicit ord: Ordering[T]): Boolean = ord.lt(o1,o2)
+
+  /**
+   * Get the map which contains the rank of all items which is currently in the sketch.
```
please add a blank line after this

pengzai6666 replied:
Added.

sscdotopen commented on src/main/scala/com/amazon/deequ/analyzers/QuantileNonSample.scala:
```diff
@@ -0,0 +1,307 @@
+
+  /**
+   * Get the map which contains the rank of all items which is currently in the sketch.
+   * @return the sorted Listmap (by key)which contains the rank of all items which is currently in the sketch (RankMap)
```
, which

pengzai6666 replied:
Updated.

sscdotopen commented on src/main/scala/com/amazon/deequ/analyzers/QuantileNonSample.scala:
```diff
@@ -0,0 +1,307 @@
+   * Get the map which contains the rank of all items which is currently in the sketch.
+   * @return the sorted Listmap (by key)which contains the rank of all items which is currently in the sketch (RankMap)
+   */
+  def getRankMap(): ListMap[T,Long] = {
```
please don't use `_1` and `_2`, destructure tuples and use descriptive variable names

pengzai6666 replied:
Use item to replace _1, and weight to replace _2

sscdotopen commented on src/main/scala/com/amazon/deequ/analyzers/QuantileNonSample.scala:
```diff
@@ -0,0 +1,307 @@
+   * @return the sorted Listmap (by key)which contains the rank of all items which is currently in the sketch (RankMap)
+   */
+  def getRankMap(): ListMap[T,Long] = {
+    val sortedOutput = ListMap(output.toSeq.sortBy(_._1):_*)
```
please don't use `_1` and `_2`, destructure tuples and use descriptive variable names

pengzai6666 replied:
Use item to replace _1, and weight to replace _2

sscdotopen commented on src/main/scala/com/amazon/deequ/analyzers/QuantileNonSample.scala:
```diff
@@ -0,0 +1,307 @@
+    val sortedOutput = ListMap(output.toSeq.sortBy(_._1):_*)
+    var states = scala.collection.mutable.Map[T,Long]()
+    var runningRank = 0L
+    sortedOutput.foreach{ tuple =>
```
please don't use `_1` and `_2`, destructure tuples and use descriptive variable names
please add a space before {

pengzai6666 replied:
Use item to replace _1, and weight to replace _2
Space added.

sscdotopen commented on src/main/scala/com/amazon/deequ/analyzers/QuantileNonSample.scala:
```diff
@@ -0,0 +1,307 @@
+    var states = scala.collection.mutable.Map[T,Long]()
+    var runningRank = 0L
+    sortedOutput.foreach{ tuple =>
+      runningRank = runningRank + tuple._2
```
please don't use `_1` and `_2`, destructure tuples and use descriptive variable names

pengzai6666 replied:
Use item to replace _1, and weight to replace _2

sscdotopen commented on src/main/scala/com/amazon/deequ/analyzers/QuantileNonSample.scala:
```diff
@@ -0,0 +1,307 @@
+  def getCDF(): Array[(T,Double)] = {
+    val rankMap = getRankMap()
+    val tmp = rankMap.keySet.toArray
+    val totalWeight = rankMap.last._2
```
please don't use `_1` and `_2`, destructure tuples and use descriptive variable names

pengzai6666 replied:
Change to:
val (_, totalWeight) = rankMap.last

sscdotopen commented on src/main/scala/com/amazon/deequ/analyzers/QuantileNonSample.scala:
```diff
@@ -0,0 +1,307 @@
+    val tmp = rankMap.keySet.toArray
+    val totalWeight = rankMap.last._2
+    var ret = ArrayBuffer[(T,Double)]()
+    tmp.foreach{ item =>
```
please add a space before {

pengzai6666 replied:
Added.

sscdotopen commented on src/main/scala/com/amazon/deequ/analyzers/QuantileNonSample.scala:
```diff
@@ -0,0 +1,307 @@
+   */
+  def getRank(item:T): Long = {
+    var r = 0L
+    output.foreach(tuple => {
```
please don't use `_1` and `_2`, destructure tuples and use descriptive variable names

pengzai6666 replied:
Updated.

sscdotopen commented on src/main/scala/com/amazon/deequ/analyzers/QuantileNonSample.scala:
```diff
@@ -0,0 +1,307 @@
+   * @return the estimated rank of the query item in sketch
+   */
+  def getRank(item:T): Long = {
+    var r = 0L
```
rank

pengzai6666 replied:
Renamed.

sscdotopen commented on src/main/scala/com/amazon/deequ/analyzers/QuantileNonSample.scala:
```diff
@@ -0,0 +1,307 @@
+   * @return the estimated rank of the query item in sketch
+   */
+  def getRankExclusive(item:T): Long = {
+    var r = 0L
```
rank

pengzai6666 replied:
Renamed.

sscdotopen commented on src/main/scala/com/amazon/deequ/analyzers/QuantileNonSample.scala:
```diff
@@ -0,0 +1,307 @@
+  def getRankExclusive(item:T): Long = {
+    var r = 0L
+    output.foreach(tuple => {
+      if (compareSmaller(tuple._1,item)) {
```
please don't use `_1` and `_2`, destructure tuples and use descriptive variable names

pengzai6666 replied:
Updated.

sscdotopen commented on src/main/scala/com/amazon/deequ/analyzers/QuantileNonSample.scala:
```diff
@@ -0,0 +1,307 @@
+    var curRank = 0L
+    breakable {
+      for (target <- rankMap) {
+        if (compareGreater(target._1, item))
```
please don't use `_1` and `_2`, destructure tuples and use descriptive variable names

pengzai6666 replied:
Updated.

sscdotopen commented on src/main/scala/com/amazon/deequ/analyzers/QuantileNonSample.scala:
```diff
@@ -0,0 +1,307 @@
+
+
+  /**
+   * merge two sketches into a single one
```
Merge

pengzai6666 replied:
Modified.

sscdotopen commented on src/main/scala/com/amazon/deequ/analyzers/QuantileNonSample.scala:
```diff
@@ -0,0 +1,307 @@
+  }
+
+  private def output: Array[(T,Long)] = {
+    compactors.toArray.slice(0,curNumOfCompactors).zipWithIndex.flatMap{case (compactor,i) =>
```
please break this into more statements

pengzai6666 replied:
Use compactorArray and compactorIndexMap to break up the statements.

sscdotopen commented on src/main/scala/com/amazon/deequ/analyzers/QuantileNonSample.scala:
```diff
@@ -0,0 +1,307 @@
+   * @return quantiles 1/q through (q-1)/q
+   */
+  def quantiles(q: Int) : Array[T] = {
+    val sortedItems = output.sortBy(_._1)
```
please don't use `_1` and `_2`, destructure tuples and use descriptive variable names

pengzai6666 replied:
Updated.

sscdotopen commented on src/main/scala/com/amazon/deequ/analyzers/QuantileNonSample.scala:
```diff
@@ -0,0 +1,307 @@
+   */
+  def quantiles(q: Int) : Array[T] = {
+    val sortedItems = output.sortBy(_._1)
+    val size = sortedItems.map(_._2).sum
```
please use descriptive variable names

pengzai6666 replied:
Change the variable "size" to "totalWeight".

sscdotopen commented on src/main/scala/com/amazon/deequ/analyzers/QuantileNonSample.scala:
```diff
@@ -0,0 +1,307 @@
+    val quantiles = Array.fill[T](q-1)(sortedItems(0)._1)
+
+
+    while (i<sortedItems.length && curq<q) {
```
please don't use `_1` and `_2`, destructure tuples and use descriptive variable names

pengzai6666 replied:
Updated.

sscdotopen commented on src/main/scala/com/amazon/deequ/analyzers/QuantileNonSample.scala:
```diff
@@ -0,0 +1,307 @@
+    quantiles
+  }
+
+  /**
```
Please format the rest of this file according to our [coding conventions](https://github.com/databricks/scala-style-guide)

pengzai6666 replied:
Added comments for public functions.

sscdotopen commented on src/main/scala/com/amazon/deequ/analyzers/StateProvider.scala:
```diff
@@ -224,15 +223,9 @@ case class HdfsStateProvider(
       identifier: String)
     : Unit = {
 
-    val saveMode = if (allowOverwrite) {
```
please rebase to the latest mainline

pengzai6666 replied:
I believe that I have rebased to the latest master, but feel free to check if the rebase is successful or not.

sscdotopen commented on src/main/scala/com/amazon/deequ/analyzers/catalyst/DeequFunctions.scala:
```diff
@@ -79,6 +80,12 @@ object DeequFunctions {
     statefulDataType(column)
   }
+
+  def stateful_kll(column: Column, sketchSize: Int = KLLSketch.defaultSketchSize,
```
Please format according to our [coding conventions](https://github.com/databricks/scala-style-guide)

pengzai6666 replied:
Updated.

sscdotopen commented on src/main/scala/com/amazon/deequ/analyzers/catalyst/KLLSketchSerializer.scala:
```diff
@@ -0,0 +1,105 @@
+package com.amazon.deequ.analyzers.catalyst
```
This file needs a license header

pengzai6666 replied:
Added.

sscdotopen commented on src/main/scala/com/amazon/deequ/analyzers/catalyst/StatefulKLLSketch.scala:
```diff
@@ -0,0 +1,91 @@
+package org.apache.spark.sql
```
Please format according to our [coding conventions](https://github.com/databricks/scala-style-guide)

pengzai6666 replied:
Added the license header.

sscdotopen commented on src/main/scala/com/amazon/deequ/analyzers/catalyst/StatefulKLLSketch.scala:
```diff
@@ -0,0 +1,91 @@
+import org.apache.spark.sql.types._
+
+
+private [sql] class StatefulKLLSketch(sketchSize: Int,
```
Please format according to our [coding conventions](https://github.com/databricks/scala-style-guide)

pengzai6666 replied:
Done.

sscdotopen commented on src/main/scala/com/amazon/deequ/analyzers/catalyst/StatefulKLLSketch.scala:
```diff
@@ -0,0 +1,91 @@
+    KLLSketchSerializer.serializer.deserialize(bytes)
+  }
+
+  def typeCheck(input: Row) : Boolean = {
```
This function is unclear to me.

pengzai6666 replied:
This function is to check if the input is null, if it is null we will break without further executing the rest of the code.

Replace the function with isNullAt() to avoid confusion.

sscdotopen commented on src/main/scala/com/amazon/deequ/examples/ExampleUtils.scala:
```diff
@@ -19,12 +19,17 @@ package com.amazon.deequ.examples
       .master("local")
       .appName("test")
-      .config("spark.ui.enabled", "false")
+      .config("spark.ui.enabled", "true")
```
please revert this

pengzai6666 replied:
Reverted.

sscdotopen commented on src/main/scala/com/amazon/deequ/examples/PengExample.scala:
```diff
@@ -0,0 +1,60 @@
+package com.amazon.deequ.examples
```
Please rename this file and give it a license header

pengzai6666 replied:
Actually I wrote this example for testing how much time it could take to run deequ checks on different scales of datasets. It is irrelevant to the KLL Sketch and I will remove this file.

sscdotopen commented on src/main/scala/com/amazon/deequ/metrics/KLLMetric.scala:
```diff
@@ -0,0 +1,53 @@
+  // find the index of bucket which has the largest count
+  def argmax: Int = {
+    var max_store = 0L
+    var ret = 0
```
Please use a different variable name

pengzai6666 replied:
Done.

sscdotopen commented on src/main/scala/com/amazon/deequ/profiles/ColumnProfile.scala:
```diff
@@ -49,6 +51,7 @@ case class NumericColumnProfile(
     isDataTypeInferred: Boolean,
     typeCounts: Map[String, Long],
     histogram: Option[Distribution],
+    KLL: Option[BucketDistribution],
```
kll

pengzai6666 replied:
Refactored.

sscdotopen commented on src/main/scala/com/amazon/deequ/profiles/ColumnProfile.scala:
```diff
@@ -121,6 +124,40 @@ object ColumnProfiles {
+            val store = new JsonObject()
+            store.add("parameters", entry)
+
+            // convert kllSketch.data to JSON String
```
please remove the comments

pengzai6666 replied:
Removed.

sscdotopen commented on src/test/scala/com/amazon/deequ/analyzers/StateProviderTest.scala:
```diff
@@ -101,50 +101,6 @@ class StateProviderTest extends WordSpec with Matchers with SparkContextSpec wit
 
       assertCorrectlyApproxQuantileState(provider, provider, ApproxQuantile("price", 0.5), data)
     }
```
why is this deleted?

pengzai6666 replied:
I haven't make changes to StateProvider.scala or  StateProviderTest.scala. This is probably because I didn't rebase the latest master. Will check again after rebase.

sscdotopen commented on src/test/scala/com/amazon/deequ/profiles/ColumnProfilerTest.scala:
```diff
@@ -96,6 +96,7 @@ class ColumnProfilerTest extends WordSpec with Matchers with SparkContextSpec
           "String" -> 0
```
Please explain some of the changes in this file

pengzai6666 replied:
Since I have added kll to the NumericalProfile, I added 'None' in Numerical Profile in some of the tests to in order to let the tests pass. The other changes are not necessary and I have reverted them back.

pengzai6666 commented:
Hi Sebastian, 

First of all, thank you so much for checking on every details of my implementation and giving precious feedback, especially on weekends!

Second, I have made relevant changes to meet the coding conventions, added license headers and rebased to the latest master. But it is possible that I may miss out/misunderstand something so I would appreciate it if you could double check on the code to make sure that it meets your requirement.

My next step would be adding serialize/deserialize functions between KLL Sketches and Json strings, as well as writing a distance computation function based on L-Infinity distance inside the KLL Sketch class. I will make new commits to the PR and let you know when it is done.

sscdotopen commented:
Hi Peng,
Thank you very much for the contributions. Could you check why the build of this paper still doesn't work? I think its because some files still miss license headers. I would be inclined to already merge this PR if its builds correctly and move the distance computations into another PR.

sscdotopen commented on src/main/scala/com/amazon/deequ/analyzers/KLLSketch.scala:
```diff
@@ -0,0 +1,158 @@
+  /** Add up states by merging sketches */
+  override def sum(other: KLLState): KLLState = {
+    val mergedSketch = qSketch.merge(other.qSketch)
+    KLLState(mergedSketch, Math.max(globalMax,other.globalMax), Math.min(globalMin,other.globalMin))
```
Please add spaces after the commas

pengzai6666 replied:
Done.

sscdotopen commented on src/main/scala/com/amazon/deequ/analyzers/KLLSketch.scala:
```diff
@@ -0,0 +1,158 @@
+}
+
+object KLLState{
+  /**
```
Please add a blank line here

pengzai6666 replied:
Done.

sscdotopen commented on src/main/scala/com/amazon/deequ/analyzers/KLLSketch.scala:
```diff
@@ -0,0 +1,158 @@
+  val DEFAULT_SKETCH_SIZE = 2
+  val DEFAULT_SHRINKING_FACTOR = 0.64
+  val MAXIMUM_ALLOWED_DETAIL_BINS = 2
+}
```
Please add a newline here

pengzai6666 replied:
Done.

sscdotopen commented on src/main/scala/com/amazon/deequ/analyzers/NonSampleCompactor.scala:
```diff
@@ -0,0 +1,53 @@
+package com.amazon.deequ.analyzers
```
This file needs a license header

pengzai6666 replied:
Added.

sscdotopen commented on src/main/scala/com/amazon/deequ/analyzers/NonSampleCompactor.scala:
```diff
@@ -0,0 +1,53 @@
+ *           over this item type
+ */
+class NonSampleCompactor[T]()
+                           (implicit ordering: Ordering[T],
```
4 spaces indentation on these 2 lines please

pengzai6666 replied:
Done.

sscdotopen commented on src/main/scala/com/amazon/deequ/analyzers/QuantileNonSample.scala:
```diff
@@ -0,0 +1,284 @@
+      while (sumSoFar < nextThresh) {
+        val (_, weight) = sortedItems(i)
+        sumSoFar += weight
+        i+=1
```
spaces around += please

pengzai6666 replied:
Done.

sscdotopen commented on src/main/scala/com/amazon/deequ/analyzers/catalyst/KLLSketchSerializer.scala:
```diff
@@ -0,0 +1,121 @@
+      var i = 0
+      while (i < compactorLength) {
+        var compactor = new NonSampleCompactor[Double]()
+        val numOfCompress= buffer.getInt()
```
space before = please

pengzai6666 replied:
Done.

sscdotopen commented on src/main/scala/com/amazon/deequ/analyzers/catalyst/StatefulKLLSketch.scala:
```diff
@@ -0,0 +1,98 @@
+  override def inputSchema: StructType = StructType(StructField("value", DoubleType) :: Nil)
+
+  override def bufferSchema: StructType = StructType(StructField("data", BinaryType) ::
+    StructField("minimum",DoubleType) :: StructField("maximum",DoubleType) :: Nil)
```
Please add spaces after the commas

pengzai6666 replied:
Done.

sscdotopen commented on src/main/scala/com/amazon/deequ/analyzers/catalyst/StatefulKLLSketch.scala:
```diff
@@ -0,0 +1,98 @@
+  override def deterministic: Boolean = true
+
+  override def initialize(buffer: MutableAggregationBuffer): Unit = {
+    val qsketch =  new QuantileNonSample[Double](sketchSize, shrinkingFactor)
```
only one space after = please

pengzai6666 replied:
Done.

sscdotopen commented on src/main/scala/com/amazon/deequ/analyzers/catalyst/StatefulKLLSketch.scala:
```diff
@@ -0,0 +1,98 @@
+  }
+
+  override def update(buffer: MutableAggregationBuffer, input: Row): Unit = {
+    if (input.isNullAt(OBJECT_POS))
```
please use braces for the if statement

pengzai6666 replied:
Done.

sscdotopen commented on src/main/scala/com/amazon/deequ/analyzers/catalyst/StatefulKLLSketch.scala:
```diff
@@ -0,0 +1,98 @@
+  }
+
+  override def merge(buffer1: MutableAggregationBuffer, buffer2: Row): Unit = {
+    if (buffer2.isNullAt(OBJECT_POS))
```
please use braces for the if statement

pengzai6666 replied:
Done.

sscdotopen commented on src/main/scala/com/amazon/deequ/analyzers/catalyst/StatefulKLLSketch.scala:
```diff
@@ -0,0 +1,98 @@
+    toBytes(buffer.getDouble(MIN_POS), buffer.getDouble(MAX_POS), buffer.getAs[Array[Byte]](OBJECT_POS))
+  }
+
+  def toBytes(min: Double, max: Double, obj:Array[Byte]): Array[Byte]= {
```
space after:

pengzai6666 replied:
Space added.

sscdotopen commented on src/main/scala/com/amazon/deequ/metrics/KLLMetric.scala:
```diff
@@ -0,0 +1,53 @@
+
+case class BucketDistribution(buckets: List[BucketValue], parameters: List[Double], data: Array[Array[Double]]) {
+
+  // Get relevant bucketValue with index of bucket
```
Please use a /**/ comment

pengzai6666 replied:
Added.

sscdotopen commented on src/main/scala/com/amazon/deequ/metrics/KLLMetric.scala:
```diff
@@ -0,0 +1,53 @@
+    buckets(key)
+  }
+
+  // find the index of bucket which has the largest count
```
Please use a /**/ comment

pengzai6666 replied:
Added.

sscdotopen commented:
There's still a couple of style violations, the travis output will give you some pointers: https://travis-ci.org/awslabs/deequ/builds/601452300

pengzai6666 commented:
Yeah exactly. I have fixed the style violations and so far the newest commit is built successfully, I think.

sscdotopen commented:
I still get a build failure locally due to some missing license file, let me have a detailed look

sscdotopen commented:
Oh that was a local file of mine ;)

pengzai6666 commented:
;)

pengzai6666 commented:
I would appreciate it if you could consider merging the PR if no more problems emerge.

sscdotopen commented:
I'm getting a test failure in https://github.com/awslabs/deequ/blob/master/src/test/scala/com/amazon/deequ/suggestions/ConstraintSuggestionRunnerTest.scala#L41 when I build your branch locally. I am currently trying to understand what the problem is

pengzai6666 commented:
I think it is because we have a data: Array[Array[Double]] inside KLLMetric, so we could not compare them directly by:
assert(expectedResult.columnProfiles == actualResult.columnProfiles)
because we are comparing to the object itself, not the contents inside it.
After serialization into JSON, we can compare to the contents by:
assert(expectedConstraintSuggestionJson == actualConstraintSuggestionJson)

sscdotopen commented:
Yes, makes sense. Could you add a custom equals implementation for the KLLMetric?

sscdotopen commented:
There are still 3 scala style violations:

```

sscdotopen commented:
There is still one error message left: 
`error file=/home/ssc/Entwicklung/projects/deequ/src/main/scala/com/amazon/deequ/profiles/ColumnProfiler.scala message=argcount.message line=91 column=21`

Btw, you can locally test this with `mvn clean install`
